server/src/main/java/org/opensearch/rest/action/document/RestBulkStreamingAction.java

```diff
@@ -0,0 +1,199 @@
+ * </pre>
+ *
+ * @opensearch.api
+ */
+public class RestBulkStreamingAction extends BaseRestHandler {
+    private static final BulkResponse EMPTY = new BulkResponse(new BulkItemResponse[0], 0L);
+    private final boolean allowExplicitIndex;
+
+    public RestBulkStreamingAction(Settings settings) {
+        this.allowExplicitIndex = MULTI_ALLOW_EXPLICIT_INDEX.get(settings);
+    }
+
+    @Override
+    public List<Route> routes() {
+        return unmodifiableList(
+            asList(
+                new Route(POST, "/_bulk/stream"),
+                new Route(PUT, "/_bulk/stream"),
+                new Route(POST, "/{index}/_bulk/stream"),
+                new Route(PUT, "/{index}/_bulk/stream")
+            )
+        );
+    }
+
+    @Override
+    public String getName() {
+        return "streaming_bulk_action";
+    }
+
+    @Override
+    public RestChannelConsumer prepareRequest(final RestRequest request, final NodeClient client) throws IOException {
+        final String defaultIndex = request.param("index");
+        final String defaultRouting = request.param("routing");
+        final String defaultPipeline = request.param("pipeline");
+        final String waitForActiveShards = request.param("wait_for_active_shards");
+        final Boolean defaultRequireAlias = request.paramAsBoolean(DocWriteRequest.REQUIRE_ALIAS, null);
+        final TimeValue timeout = request.paramAsTime("timeout", BulkShardRequest.DEFAULT_TIMEOUT);
+        final String refresh = request.param("refresh");
+
+        final StreamingRestChannelConsumer consumer = (channel) -> {
+            final MediaType mediaType = request.getMediaType();
+
+            // Set the content type and the status code before sending the response stream over
+            channel.prepareResponse(RestStatus.OK, Map.of("Content-Type", List.of(mediaType.mediaTypeWithoutParameters())));
+
+            // This is initial implementation at the moment which transforms each single request stream chunk into
+            // individual bulk request and streams each response back. Another source of inefficiency comes from converting
+            // bulk response from raw (json/yaml/...) to model and back to raw (json/yaml/...).
+
+            // TODOs:
+            // - add batching (by interval and/or count)
+            // - eliminate serialization inefficiencies
+            Flux.from(channel).map(chunk -> {
+                FetchSourceContext defaultFetchSourceContext = FetchSourceContext.parseFromRestRequest(request);
+                BulkRequest bulkRequest = Requests.bulkRequest();
+                if (waitForActiveShards != null) {
+                    bulkRequest.waitForActiveShards(ActiveShardCount.parseString(waitForActiveShards));
+                }
+
+                bulkRequest.timeout(timeout);
+                bulkRequest.setRefreshPolicy(refresh);
+
+                try {
+                    bulkRequest.add(
+                        chunk.content(),
```

Comment:
How does this work that an individual chunk here is guaranteed to contain enough data to form a bulk request?

Reply:
For example, if I send the following:

```
{ "create": { "_index": "movies", "_id": "tt1392214" } }
{ "title": "Prisoners", "year": 2013 }
```

then the server must receive both JSON documents in order to form a valid bulk request. If my client is just writing those to the stream on the other side of the connection, how is it guaranteed that the "chunks" processed here don't separate these two documents as distinct chunks? (there may be a lot I don't understand about how this works...)

Reply:
So the chunking is controlled by client, that would be a single chunk for server:

```
{ "create": { "_index": "movies", "_id": "tt1392214" } }
{ "title": "Prisoners", "year": 2013 }
```

Reply:
I see, I think I need to learn a little more about how HTTP/2 works here :)